Musing on microservices (#2)
Ottermad/ottermad.github.io: _posts/2021-08-02-Musing-on-Microservices.md

---
layout: post
title: Musing on Microservices
author: Charles Thomas
---

> Microservices are commonplace among technology companies of all sizes and in this post, I hope to explore some of the pros and cons of them and what you should consider when thinking about whether or not they are a good solution for the problems you're solving.

Reviewer: ```suggestion
Microservices are commonplace among technology companies of all sizes and in this post, I hope to explore some of the pros and cons of them and what you should consider when deciding if they're a good solution for the problems you're solving.
```

Let's start by quickly addressing a common question: what are microservices? Microservices are a way of structuring your application so that instead of having one big application (referred to as a monolith), you have multiple smaller applications that communicate with each other over a network.

> So why might you want to do this? Generally, microservices set out to solve two problems: scalability (how can I make my application handle more load?) and availability (how can I keep my application serving requests more of the time?). Let's deal with each of these issues separately.

Reviewer: There are definitely other problems they set out to solve:

- Individual deployability (isolating scope of bad changes, decreasing deployment surface area, easier rollbacks etc.)
- Abstraction (expose behaviour via a common protocol, developers are free to use any language / framework etc. as they please)

Probably a load of others I'm skipping

Author: I would argue that individual deployability isn't really something that comes with microservices. It comes with having good build tools and deploying frequently. You can deploy small changes and roll back quickly with most monoliths. If anything rolling back is much harder with microservices and you have to ensure compatibility between all service versions

> Microservices can help improve scalability by allowing you to run multiple copies of your app on different machines. However, you can do this with a monolith as well - just run more copies of the whole monolith  Instead, microservices provide a more subtle advantage: they allow you to scale different parts of your application separately according to demand. For example, you might have a part of your application such as authorization that has to process a very high number of requests while another part of your application such as settings has much lower traffic. In a microservices architecture, you can scale these parts of your application separately.

Reviewer: nit: Not _necessarily_ on different machines

> In principle, this is great - it allows you to scale parts of your application independently, saving you money on hosting costs and allowing you to handle higher traffic loads more easily. However, the important question to ask yourself is whether or not different parts of your application actually have drastically different scalability needs and does the ability to scale them separately outweigh the cost of the added complexity?

Reviewer: And also are you operating at a scale where the cost difference is actually significant?

> It is also worth asking what is currently limiting your ability to scale your application as a monolith? If it is CPU usage or memory, then perhaps, microservices may be a solution, however, you may be being bottlenecked by something else such as your database. In this case, microservices will not solve your problem, although the common pattern of having one database per microservice might help you. However, you can get the same effect by having your monolith connect to multiple databases.

Reviewer: ```suggestion
It is also worth asking what's currently limiting your ability to scale your monolith application? If it's CPU usage or memory, then perhaps, microservices may be a solution, however, you may be being bottlenecked by something else such as your database. In this case, microservices will not solve your problem, although the common pattern of having one database per microservice might help you. However, you can get the same effect by having your monolith connect to multiple databases.
```

Reviewer: You also tend to [comma splice](https://examples.yourdictionary.com/comma-splice-examples.html#:~:text=A%20comma%20splice%20is%20a%20common%20grammatical%20error%20in%20English.&text=A%20comma%20splice%20is%20when,to%20identify%20an%20independent%20clause.) quite a bit, leading to long sentences that can be hard to parse

It's also important to think about compound failure when dealing with microservices. With a monolith, if at least one copy of your monolith is up 99.9% of the time (ignoring external dependencies such as databases) then your uptime is 99.9%. Now, imagine you have a microservices architecture with two services A and B which must both be up for the user to do anything. If both of these services are up 99.9% of the time then your total uptime is actually 0.999 x 0.999 x 100 = 99.8%. This difference may seem small but given that these compounds for each dependency you have, you can quickly have much lower availability than you thought. 

> The other thing to consider when thinking about availability is failure modes. In a microservices architecture, you have more individual components, which means you have more combinations of things that can go wrong. This often makes it harder to reason about behaviour in the case of a bug or a crash. For example, if one service (service A) goes down but another (service B) stays up, on one hand, you can keep serving requests to service B, however, service A and B may now be out of sync with each other (for example, service B has consumed some event that service A hasn't) which now may lead to unexpected and confusing behaviour. In many cases, it is actually better for a failure to mean that your application just stops, rather than half processing requests.

Reviewer: Fault tolerance is a massive topic too, maybe you could link out to some extra reading here?

> The final thing to think about with microservices is what I'm calling: microservices as a design pattern. In some companies, microservices have become a way of managing code complexity. In some codebases where you might have a class, in others, you may now have a microservice. Now while this approach can work, if you are considering adopting microservices, for this reason, I would strongly suggest you consider whether or not it is worth the overhead of having to support such a high number of microservices.

Reviewer: Could you talk about alternatives here?

e.g. more effective abstraction (interfaces? better docs?), separating out code ownership by package?
